lib/vagrant_cloud/client.rb

```diff
@@ -1,74 +1,454 @@
+    # Default instrumentor
+    DEFAULT_INSTRUMENTOR = Instrumentor::Collection.new
 
-      @access_token = access_token
+    # @return [Instrumentor::Collection]
+    def self.instrumentor
+      DEFAULT_INSTRUMENTOR
     end
 
-    # @param [String] method
-    # @param [String] path
-    # @param [Hash] params
-    # @param [String] token
-    # @return [Hash]
-    def request(method, path, params = {}, token = nil)
-      headers = {}
+    # @return [String] Access token for Vagrant Cloud
+    attr_reader :access_token
+    # @return [String] Base request path
+    attr_reader :path_base
+    # @return [String] URL for initializing connection
+    attr_reader :url_base
+    # @return [Integer] Number of retries on idempotent requests
+    attr_reader :retry_count
+    # @return [Integer] Number of seconds to wait between requests
+    attr_reader :retry_interval
+    # @return [Instrumentor::Collection] Instrumentor in use
+    attr_reader :instrumentor
 
-      if token
-        headers['Authorization'] = "Bearer #{token}"
-      elsif @access_token
-        headers['Authorization'] = "Bearer #{@access_token}"
+    # Create a new Client instance
+    #
+    # @param [String] access_token Authentication token for API requests
+    # @param [String] url_base URL used to make API requests
+    # @param [Integer] retry_count Number of retries on idempotent requests
+    # @param [Integer] retry_interval Number of seconds to wait between requests
+    # @param [Instrumentor::Core] instrumentor Instrumentor to use
+    # @return [Client]
+    def initialize(access_token: nil, url_base: nil, retry_count: nil, retry_interval: nil, instrumentor: nil)
+      url_base = DEFAULT_URL if url_base.nil?
+      remote_url = URI.parse(url_base)
+      @url_base = "#{remote_url.scheme}://#{remote_url.host}"
+      @path_base = remote_url.path
+      @access_token = access_token.dup.freeze if access_token
+      if !@access_token && ENV["VAGRANT_CLOUD_TOKEN"]
+        @access_token = ENV["VAGRANT_CLOUD_TOKEN"].dup.freeze
+      end
+      @retry_count = retry_count.nil? ? IDEMPOTENT_RETRIES : retry_count.to_i
+      @retry_interval = retry_interval.nil? ? IDEMPOTENT_RETRY_INTERVAL : retry_interval.to_i
+      @instrumentor = instrumentor.nil? ? Instrumentor::Collection.new : instrumentor
+      headers = {}.tap do |h|
+        h["Accept"] = "application/json"
+        h["Authorization"] = "Bearer #{@access_token}" if @access_token
+        h["Content-Type"] = "application/json"
       end
+      @connection_lock = Mutex.new
+      @connection = Excon.new(url_base,
+        persistent: true,
+        headers: headers,
+        instrumentor: @instrumentor
+      )
+    end
 
-      headers['Accept'] = 'application/json'
+    # Use the remote connection
+    #
+    # @param [Boolean] wait Wait for the connection to be available
+    # @yieldparam [Excon::Connection]
+    # @return [Object]
+    def with_connection(wait: true)
```

Comment:
Does `&block` not need to be explicitly passed in?

Reply:
No, because we do not need to capture the block in a variable to do anything with it. We can just simply `yield` and Ruby will use the block that was passed.